pkg/api/mlflow/dao/repositories/namespace_cached.go

```diff
@@ -0,0 +1,186 @@
+	namespaceRepository NamespaceRepositoryProvider
+}
+
+// NewNamespaceCachedRepository creates new instance of cached repository to work with `namespace` entity.
+func NewNamespaceCachedRepository(
+	db *gorm.DB, listener dao.EventListenerProvider, namespaceRepository NamespaceRepositoryProvider,
+) (*NamespaceCachedRepository, error) {
+	cache, err := lru.New[string, *models.Namespace](1000)
+	if err != nil {
+		return nil, eris.Wrap(err, "error creating lru cache for namespace entities")
+	}
+
+	// pre load all namespaces into cache.
+	var namespaces []models.Namespace
+	if err := db.Find(&namespaces).Error; err != nil {
+		return nil, eris.Wrapf(err, "error getting namespaces")
+	}
+	//nolint:gosec
+	for _, namespace := range namespaces {
+		cache.Add(namespace.Code, &namespace)
+	}
+
+	repository := NamespaceCachedRepository{
+		db:                  db,
+		cache:               cache,
+		listener:            listener,
+		namespaceRepository: namespaceRepository,
+	}
+
+	go func() {
+		for data := range listener.Listen() {
+			if err := repository.processEvent(data); err != nil {
+				log.Errorf(`error processing incoming event: %s, error: %+v`, data, err)
+			}
+		}
+	}()
+	return &repository, nil
+}
+
+// Create creates new models.Namespace entity.
+func (r NamespaceCachedRepository) Create(ctx context.Context, namespace *models.Namespace) error {
+	if err := r.namespaceRepository.Create(ctx, namespace); err != nil {
+		return eris.Wrap(err, "error creating cached namespace entity")
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to create record in theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionCreated, namespace); err != nil {
+		return eris.Wrap(err, "error sending database event")
+	}
+	return nil
+}
+
+// Update updates existing models.Namespace entity.
+func (r NamespaceCachedRepository) Update(ctx context.Context, namespace *models.Namespace) error {
+	if err := r.namespaceRepository.Update(ctx, namespace); err != nil {
+		return eris.Wrap(err, "error updating cached namespace entity")
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to update record in theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionUpdated, namespace); err != nil {
+		return eris.Wrap(err, "error sending database event")
+	}
+	return nil
+}
+
+// GetByCode returns namespace by its Code.
+func (r NamespaceCachedRepository) GetByCode(ctx context.Context, code string) (*models.Namespace, error) {
+	result, ok := r.cache.Get(code)
+	if ok {
+		return result, nil
+	}
+
+	namespace, err := r.namespaceRepository.GetByCode(ctx, code)
+	if err != nil {
+		return nil, eris.Wrapf(err, "error getting cached namespace by code: %s", code)
+	}
+
+	// trigger database event to notify current instance and
+	// other instances to add record to theirs local cache.
+	if err := r.sendEvent(NamespaceEventActionFetched, namespace); err != nil {
```

Comment:
maybe we don't need this event publish for "Get"? the other instances will update anyway when they try to fetch, or if they have it already it is unchanged.

Reply:
i see later that your are using same notifications for local cache as well as remote -- so probably nevermind :-)